services/brig/src/Brig/API/Public.hs

```diff
@@ -1074,23 +1085,69 @@ customerExtensionCheckBlockedDomains email = do
+  if localDomain == otherDomain
+    then createLocalConnection self conn (Public.ConnectionRequest otherUser (unsafeRange "_"))
+    else throwM federationNotImplemented
+
+updateLocalConnection :: UserId -> ConnId -> UserId -> Public.ConnectionUpdate -> Handler (Public.UpdateResult Public.UserConnection)
+updateLocalConnection self conn other update = do
   let newStatus = Public.cuStatus update
   mc <- API.updateConnection self other newStatus (Just conn) !>> connError
   return $ maybe Public.Unchanged Public.Updated mc
 
-listConnections :: UserId -> Maybe UserId -> Maybe (Range 1 500 Int32) -> Handler Public.UserConnectionList
-listConnections uid start msize = do
+-- | FUTUREWORK: also update remote connections: https://wearezeta.atlassian.net/browse/SQCORE-959
+updateConnection :: UserId -> ConnId -> Qualified UserId -> Public.ConnectionUpdate -> Handler (Public.UpdateResult Public.UserConnection)
+updateConnection self conn (Qualified otherUid otherDomain) update = do
+  localDomain <- viewFederationDomain
+  if localDomain == otherDomain
+    then updateLocalConnection self conn otherUid update
+    else throwM federationNotImplemented
+
+listLocalConnections :: UserId -> Maybe UserId -> Maybe (Range 1 500 Int32) -> Handler Public.UserConnectionList
+listLocalConnections uid start msize = do
   let defaultSize = toRange (Proxy @100)
   lift $ API.lookupConnections uid start (fromMaybe defaultSize msize)
 
-getConnection :: UserId -> UserId -> Handler (Maybe Public.UserConnection)
-getConnection uid uid' = lift $ API.lookupConnection uid uid'
+-- | FUTUREWORK: also list remote connections: https://wearezeta.atlassian.net/browse/SQCORE-963
+listConnections :: UserId -> Public.ListConnectionsRequestPaginated -> Handler Public.ConnectionsPage
+listConnections uid req = do
+  localDomain <- viewFederationDomain
+  let size = Public.gmtprSize req
+  res :: C.PageWithState Data.LocalConnection <- Data.lookupLocalConnectionsPage uid convertedState (rcast size)
+  return (pageToConnectionsPage localDomain Public.PagingLocals res)
+  where
+    pageToConnectionsPage :: Domain -> Public.LocalOrRemoteTable -> Data.PageWithState Data.LocalConnection -> Public.ConnectionsPage
+    pageToConnectionsPage localDomain table page@Data.PageWithState {..} =
+      Public.MultiTablePage
+        { mtpResults = Data.localToUserConn localDomain <$> pwsResults,
+          mtpHasMore = C.pwsHasMore page,
+          -- FUTUREWORK confusingly, using 'ConversationPagingState' instead of 'ConnectionPagingState' doesn't fail any tests.
+          -- Is this type actually useless? Or the tests not good enough?
+          mtpPagingState = Public.ConnectionPagingState table (LBS.toStrict . C.unPagingState <$> pwsState)
+        }
+    mkState :: ByteString -> C.PagingState
+    mkState = C.PagingState . LBS.fromStrict
+
+    convertedState :: Maybe C.PagingState
+    convertedState = fmap mkState . Public.mtpsState =<< Public.gmtprState req
```

Comment:
I guess I haven't learned much yet from @pcapriotti spotting an error in my first attempt at threading the conversation paging state, so I can't say if this looks good. Paolo, can you take a look?